### ISSUE TITLE
Priority support

### ISSUE THREAD

**Issue by yfuruyama** [cloudspannerecosystem/spanner-cli#109]
## Intro

Recently Cloud Spanner [released](https://cloud.google.com/spanner/docs/release-notes?hl=en#March_31_2021) the explicit [priority](https://cloud.google.com/spanner/docs/cpu-utilization?hl=en#task-priority) support for data requests to Cloud Spanner.

In terms of supporting priority in spanner-cli, there would be two things we need to consider:

1. Default priority
2. Allow users to specify the priority

## Default priority

Without specifying the priority Cloud Spanner runs a request with [HIGH](https://github.com/googleapis/googleapis/blob/master/google/spanner/v1/spanner.proto#L447-L448) priority, so the current requests from spanner-cli are running with HIGH priority.

spanner-cli, however, is not meant to be used for the situation where requests should be done earlier than end-user's requests, so it might be a good idea to use MEDIUM or LOW priority as default so that it won't disrupt the end-user's requests.

## Allow users to specify the priority

To allow users to specify the priority we can think of two options.

1. Session level
2. Statement level

Session level can be achieved with the command option like `--priority=LOW` when starting up the CLI.

Statement level would be a little bit controversial as there is no query hint for priority. One idea would be treating a special comment as a priority hint (e.g. `SELECT * FROM t # priority: LOW`), but it doesn't look good.

As a reference, [TiDB](https://docs.pingcap.com/tidb/dev/sql-faq) uses `select HIGH_PRIORITY | LOW_PRIORITY * from t` notation to specify the priority.

**Comment by apstndb** [comment]
> Default priority

I agree to use non-HIGH priority as a default.
I think MEDIUM priority is better because LOW priority is lower than many system tasks and it can be deferred long time.

> Allow users to specify the priority

I think, I will commonly use the session level option because I don't want to rewrite a query text if both are implemented.

Another option may emulate transaction level priority by enhancing spanner-cli's `BEGIN [RO]`/`BEGIN RW` statements.

* Pros
  * Operations in a transaction usually use same priority.
* Cons
  * Priority of single query can't be specified unless explicit transaction.

As a reference, CRDB use `BEGIN PRIORITY HIGH`.
https://www.cockroachlabs.com/docs/v20.2/transactions#transaction-priorities

**Comment by yfuruyama** [comment]
@apstndb Thank you for comments!

> I think MEDIUM priority is better because LOW priority is lower than most of system tasks and it can be deferred long time.

+1 for MEDIUM.

> Operations in a transaction usually use same priority.

I think this assumption is fairly reasonable. Cloud Spanner allows users to use different priorities in a transaction, but I don't come up with any good use case for that when using spanner-cli.

Also it looks like the transaction level priority covers current feature set of spanner-cli. The following table lists up the Request object and corresponding RPCs where [RequestOptions (priority)](https://github.com/googleapis/googleapis/blob/master/google/spanner/v1/spanner.proto#L429-L483) can be set.

| Request object | RPC | spanner-cli notation | comment |
|----------------| --- |-------------------- | ---- |
| ExecuteSqlRequest | ExecuteSql | N/A | |
| ExecuteSqlRequest | ExecuteStreamingSql | `SELECT\|UPDATE\|DELETE ...` | |
| ExecuteBatchDmlRequest | ExecuteBatchDml | N/A | |
| ReadRequest | Read | N/A | |
| ReadRequest | StreamingRead | N/A | |
| BeginTransactionRequest | BeginTransaction | `BEGIN (RO\|RW)` | Priority is not effective |
| CommitRequest | Commit | `COMMIT` | |

So basically it's enough to support priority for ExecuteStreamingSql and Commit RPC.

> As a reference, CRDB uses BEGIN PRIORITY <LOW | NORMAL | HIGH> syntax.
> https://www.cockroachlabs.com/docs/v20.2/transactions#transaction-priorities

Thanks, this is a great reference.

Considering current `BEGIN` statement notation, we can use something like this?

```
BEGIN (RO|RW) (seconds|timestamp) (PRIORITY HIGH|MEDIUM|LOW);
```

For example,

```
BEGIN PRIORITY LOW; # RW transaction with low priority
BEGIN RO PRIORITY LOW; # RO transaction with low priority
BEGIN RO 60 PRIORITY MEDIUM; # RO transaction with 60s stale read and medium priority
BEGIN RO 2021-04-01T23:47:44+00:00 PRIORITY MEDIUM; # RO transaction with exact timestamp and medium priority
```

**Comment by apstndb** [comment]
> `BEGIN (RO|RW) (seconds|timestamp) (PRIORITY HIGH|MEDIUM|LOW);`

+1
